pkg/openstack/common.go

```diff
@@ -27,3 +39,214 @@ func EnsureDeleted(ctx context.Context, helper *helper.Helper, obj client.Object
+			for _, r := range routes.Items {
+				instanceRef := metav1.OwnerReference{
+					APIVersion:         instance.APIVersion,
+					Kind:               instance.Kind,
+					Name:               instance.GetName(),
+					UID:                instance.GetUID(),
+					BlockOwnerDeletion: ptr.To(true),
+					Controller:         ptr.To(true),
+				}
+
+				owner := metav1.GetControllerOf(&r.ObjectMeta)
+
+				// Delete the route if the service was changed not to expose a route
+				if svc.ObjectMeta.Annotations[service.AnnotationIngressCreateKey] == "false" &&
+					r.Spec.To.Name == svc.Name &&
+					owner != nil && owner.UID == instance.GetUID() {
+					// Delete any other owner refs from ref list to not block deletion until owners are gone
+					r.SetOwnerReferences([]metav1.OwnerReference{instanceRef})
+
+					// Delete route
+					err := helper.GetClient().Delete(ctx, &r)
+					if err != nil && !k8s_errors.IsNotFound(err) {
+						err = fmt.Errorf("Error deleting route %s: %w", r.Name, err)
+						return svcOverrides, ctrl.Result{}, err
+					}
+
+					if svcOverride.EndpointURL != nil {
+						svcOverride.EndpointURL = nil
+						helper.GetLogger().Info(fmt.Sprintf("Service %s override endpointURL removed", svc.Name))
+					}
+				}
+			}
+		}
+
+		// If the service has the create ingress annotation and its a default ClusterIP service -> create route
+		if svc.ObjectMeta.Annotations[service.AnnotationIngressCreateKey] == "true" && svc.Spec.Type == k8s_corev1.ServiceTypeClusterIP {
+			if instance.Status.Conditions.Get(condType) == nil {
+				instance.Status.Conditions.Set(condition.UnknownCondition(
+					condType,
+					condition.InitReason,
+					corev1.OpenStackControlPlaneExposeServiceReadyInitMessage,
+					owner.GetName(),
+					svc.Name,
+				))
+			}
+
+			if svcOverride.EmbeddedLabelsAnnotations == nil {
+				svcOverride.EmbeddedLabelsAnnotations = &service.EmbeddedLabelsAnnotations{}
+			}
+
+			if labelVal, ok := svcOverride.EmbeddedLabelsAnnotations.Labels[common.AppSelector]; ok {
+				rd.ServiceLabel = map[string]string{common.AppSelector: labelVal}
+				rd.RouteName = labelVal
+			}
+
+			ctrlResult, err := rd.CreateRoute(ctx, helper, owner)
+			if err != nil {
+				instance.Status.Conditions.Set(condition.FalseCondition(
+					condType,
+					condition.ErrorReason,
+					condition.SeverityWarning,
+					corev1.OpenStackControlPlaneExposeServiceReadyErrorMessage,
+					owner.GetName(),
+					rd.RouteName,
+					err.Error()))
+
+				return svcOverrides, ctrlResult, err
+			} else if (ctrlResult != ctrl.Result{}) {
+				return svcOverrides, ctrlResult, nil
+			}
+
+			cleanCondition = true
+
+			// update override for the service with the route endpoint url
+			if rd.endpointURL != "" {
+				// Any trailing path will be added on the service-operator level.
+				svcOverride.EndpointURL = &rd.endpointURL
+				instance.Status.Conditions.MarkTrue(condType, corev1.OpenStackControlPlaneExposeServiceReadyMessage, owner.GetName())
```

Comment:
Will this condition gets cleared at L211 as we set cleanCondition at L197 after the Route was created?

Reply:
see above